File: libraries/C/trainer/trainer.c
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdbool.h>

#include "civetweb.h"
#include "cJSON.h"

#include "harness.h"


const bool CHATTY_SERVER = false;

static size_t _reserveMemSize = 0;

static int setup_handler(struct mg_connection *conn, void *cbdata) {
	if (CHATTY_SERVER) {
		printf("/setup");
	}

	char buffer[4096];
	int buffer_len = mg_read(conn, buffer, sizeof(buffer) - 1);
	buffer[buffer_len] = 0;
	if (CHATTY_SERVER) {
		printf(" w/%s\n", buffer);
	}

	cJSON *obj, *elem;
	obj = cJSON_Parse(buffer);
	if (obj == NULL) {
		mg_send_http_error(conn, 400, "Could not parse JSON");
		return 400;
	}

	elem = cJSON_GetObjectItem(obj, "reserve");
	if (!cJSON_IsNumber(elem)) {
		cJSON_Delete(obj);
		mg_send_http_error(conn, 400, "No reserve size given in JSON");
		return 400;
	}
	_reserveMemSize = elem->valueint;
	cJSON_Delete(obj);

	const uint8_t* reserveBlockPtr = (uint8_t* )simulateSetup(_reserveMemSize);
	if (reserveBlockPtr == NULL) {
		mg_send_http_error(conn, 500, "Could not reserve memory");
		return 500;
	}

	mg_send_http_ok(conn, "application/octet-stream", _reserveMemSize);
	mg_write(conn, reserveBlockPtr, _reserveMemSize);
	return 0;
}

int send_json_mem_response(struct mg_connection *conn, int code, bool success, char* msg) {
	mg_printf(conn, "HTTP/1.1 %d %s\r\n", code, mg_get_response_code_text(conn, code));
	mg_printf(conn, "Content-Type: application/json;\r\n");
	mg_printf(conn, "\r\n");
	mg_printf(conn, "{\"success\": %s, \"mem\": \"%s\"}", success ? "true" : "false", msg);
	return code;
}

static int receiveGameParams_handler(struct mg_connection *conn, void *cbdata) {
	if (CHATTY_SERVER) {
		printf("/receiveGameParams");
	}

	char buffer[4096];
	int buffer_len = mg_read(conn, buffer, sizeof(buffer) - 1);
	buffer[buffer_len] = 0;
	if (CHATTY_SERVER) {
		printf(" w/%s\n", buffer);
	}

	cJSON *obj, *elem;
	obj = cJSON_Parse(buffer);
	if (obj == NULL) {
		return send_json_mem_response(conn, 400, false, "Could not parse JSON");
	}

	elem = cJSON_GetObjectItem(obj, "offset");
	if (!cJSON_IsNumber(elem)) {
		cJSON_Delete(obj);
		return send_json_mem_response(conn, 400, false, "No offset given in JSON");
	}
	size_t offset = elem->valueint;

	elem = cJSON_GetObjectItem(obj, "mem");
	if (!cJSON_IsString(elem)) {
		cJSON_Delete(obj);
		return send_json_mem_response(conn, 400, false, "No memory string given in JSON");
	}
	char* encodedMem = elem->valuestring;
	size_t encodedMemLen = strlen(encodedMem);

	size_t decodedBlockLen = _reserveMemSize + 1;
	uint8_t* decodedBuffer = (uint8_t*)malloc(decodedBlockLen);
	int res = mg_base64_decode(encodedMem, encodedMemLen, decodedBuffer, &decodedBlockLen);
	if (res != -1 || decodedBlockLen - 1 != _reserveMemSize) {
		cJSON_Delete(obj);
		free(decodedBuffer);
		return send_json_mem_response(conn, 400, false, "Could not decode passed memory string");
	}
	cJSON_Delete(obj);

	uint8_t* updatedBuffer = (uint8_t*)simulateReceiveGameParams(decodedBuffer, offset);
	free(decodedBuffer);

	if (updatedBuffer == NULL) {
		return send_json_mem_response(conn, 400, false, "Bot declined game parameters");
	}

	size_t encodedBufferLen = ((_reserveMemSize + 2) / 3) * 4 + 1;
	char* encodedBuffer = malloc(encodedBufferLen);
	res = mg_base64_encode((const unsigned char*)updatedBuffer, _reserveMemSize, encodedBuffer, &encodedBufferLen);
	if (res != -1) {
		free(encodedBuffer);
		return send_json_mem_response(conn, 500, false, "Could not encode reserve memory");
	}

	send_json_mem_response(conn, 200, true, encodedBuffer);

	free(encodedBuffer);
	return 200;
}

static int tick_handler(struct mg_connection *conn, void *cbdata) {
	if (CHATTY_SERVER) {
		printf("/tick");
	}

	char buffer[4096];
	int buffer_len = mg_read(conn, buffer, sizeof(buffer) - 1);
	buffer[buffer_len] = 0;
	if (CHATTY_SERVER) {
		printf(" w/%s\n", buffer);
	}

	cJSON *obj, *elem;
	obj = cJSON_Parse(buffer);
	if (obj == NULL) {
		return send_json_mem_response(conn, 400, false, "Could not parse JSON");
	}

	elem = cJSON_GetObjectItem(obj, "offset");
	if (!cJSON_IsNumber(elem)) {
		cJSON_Delete(obj);
		return send_json_mem_response(conn, 400, false, "No offset given in JSON");
	}
	size_t offset = elem->valueint;

	elem = cJSON_GetObjectItem(obj, "mem");
	if (!cJSON_IsString(elem)) {
		cJSON_Delete(obj);
		return send_json_mem_response(conn, 400, false, "No memory string given in JSON");
	}
	char* encodedMem = elem->valuestring;
	size_t encodedMemLen = strlen(encodedMem);

	size_t decodedBlockLen = _reserveMemSize + 1;
	uint8_t* decodedBuffer = (uint8_t*)malloc(decodedBlockLen);
	int res = mg_base64_decode(encodedMem, encodedMemLen, decodedBuffer, &decodedBlockLen);
	if (res != -1 || decodedBlockLen - 1 != _reserveMemSize) {
		cJSON_Delete(obj);
		free(decodedBuffer);
		return send_json_mem_response(conn, 400, false, "Could not decode passed memory string");
	}
	cJSON_Delete(obj);

	uint8_t* updatedBuffer = (uint8_t*)simulateTick(decodedBuffer, offset);
	free(decodedBuffer);

	size_t encodedBufferLen = ((_reserveMemSize + 2) / 3) * 4 + 1;
	char* encodedBuffer = malloc(encodedBufferLen);
	res = mg_base64_encode((const unsigned char*)updatedBuffer, _reserveMemSize, encodedBuffer, &encodedBufferLen);
	if (res != -1) {
		free(encodedBuffer);
		return send_json_mem_response(conn, 500, false, "Could not encode reserve memory");
	}

	send_json_mem_response(conn, 200, true, encodedBuffer);

	free(encodedBuffer);
	return 200;
}


int main(void) {
	struct mg_context *ctx;
	struct mg_callbacks callbacks;
	const char *options[] = {
		"listening_ports", "9090",
		"num_threads", "1",
		NULL
	};

	memset(&callbacks, 0, sizeof(callbacks));

	ctx = mg_start(&callbacks, 0, options);
	if (!ctx) {
		printf("Cannot start trainer :(\n");
		return 1;
	}

	mg_set_request_handler(ctx, "/setup", setup_handler, NULL);
	mg_set_request_handler(ctx, "/receiveGameParams", receiveGameParams_handler, NULL);
	mg_set_request_handler(ctx, "/tick", tick_handler, NULL);

	printf("Listening...\n\n");

	while (1) {
		sleep(1);
	}

	mg_stop(ctx);
	return 0;
}

File: libraries/Go/trainer/trainer.go
package wasmbot_trainer

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"unsafe"
)

const CHATTY_SERVER bool = true

type BotImplementation struct {
	ClientInitializeFunc     func()
	SetupFunc                func(uintptr) uintptr
	ReceiveGameParamsFunc    func(uintptr) bool
	TickFunc                 func(uintptr)
	HostLogFunction          *func(int32, uintptr, uintptr)
	HostGetRandomIntFunction *func(int32, int32) int32
}

type SetupRequest struct {
	Reserve int `json:"reserve"`
}

type MemoryRequest struct {
	Offset int    `json:"offset"`
	Mem    string `json:"mem"`
}

type MemoryResult struct {
	Success bool   `json:"success"`
	Mem     string `json:"mem"`
}

var botImpl BotImplementation

func TrainerMain(impl BotImplementation) {
	botImpl = impl

	setupHarness()

	http.HandleFunc("/setup", setupHandler)
	http.HandleFunc("/receiveGameParams", receiveGameParamsHandler)
	http.HandleFunc("/tick", tickHandler)

	fmt.Printf("Listening...\n\n")
	err := http.ListenAndServe(":9090", nil)
	if err != nil {
		os.Exit(1)
	}
}

func setupHandler(w http.ResponseWriter, r *http.Request) {
	if CHATTY_SERVER {
		fmt.Print("/setup w/")
	}
	body, err := io.ReadAll(r.Body)
	if err != nil {
		errMsg := fmt.Errorf("Error reading body: %w", err)
		fmt.Fprintf(os.Stderr, errMsg.Error())
		return
	}
	if CHATTY_SERVER {
		fmt.Printf("%s\n", body)
	}

	var setup SetupRequest
	err = json.Unmarshal(body, &setup)
	if err != nil {
		errMsg := fmt.Errorf("Error parsing JSON: %w", err)
		fmt.Fprintf(os.Stderr, errMsg.Error())
		return
	}

	reserve := simulateSetup(uintptr(setup.Reserve))

	if reserve == 0 {
		w.WriteHeader(http.StatusInternalServerError)
		io.WriteString(w, "Could not allocate reserve block")
		return
	}

	if CHATTY_SERVER {
		fmt.Println("  [SUCCESS!]")
	}

	reserveBlock := unsafe.Slice((*byte)(unsafe.Pointer(reserve)), setup.Reserve)
	w.Header().Add("Content-Type", "application/octet-stream")
	w.Write(reserveBlock)
}

func sendJsonMemResponse(w http.ResponseWriter, code int, success bool, msg string) {
	data := MemoryResult{
		Success: success,
		Mem:     msg,
	}
	body, err := json.Marshal(data)
	if err != nil {
		errMsg := fmt.Errorf("Could not marshal data: %w", err)
		fmt.Fprintf(os.Stderr, errMsg.Error())
		w.WriteHeader(500)
		return
	}
	w.WriteHeader(code)
	w.Header().Add("Content-Type", "application/json")
	w.Write(body)
}

func receiveGameParamsHandler(w http.ResponseWriter, r *http.Request) {
	if CHATTY_SERVER {
		fmt.Print("/receiveGameParams w/")
	}
	body, err := io.ReadAll(r.Body)
	if err != nil {
		errMsg := fmt.Errorf("Error reading body: %w", err)
		sendJsonMemResponse(w, 400, false, errMsg.Error())
		return
	}
	if CHATTY_SERVER {
		fmt.Printf("%s\n", body)
	}

	var memReq MemoryRequest
	err = json.Unmarshal(body, &memReq)
	if err != nil {
		errMsg := fmt.Errorf("Error parsing JSON: %w", err)
		sendJsonMemResponse(w, 400, false, errMsg.Error())
		return
	}

	decodedBuffer, err := base64.StdEncoding.DecodeString(memReq.Mem)
	if err != nil {
		errMsg := fmt.Errorf("Error decoding memory string: %w", err)
		sendJsonMemResponse(w, 400, false, errMsg.Error())
		return
	}

	updatedBuffer := simulateReceiveGameParams(decodedBuffer, uintptr(memReq.Offset))

	if updatedBuffer == nil {
		sendJsonMemResponse(w, 400, false, "Bot declined game parameters")
		return
	}

	encodedMem := base64.StdEncoding.EncodeToString(updatedBuffer)
	sendJsonMemResponse(w, 200, true, encodedMem)
}

func tickHandler(w http.ResponseWriter, r *http.Request) {
	if CHATTY_SERVER {
		fmt.Print("/tick w/")
	}
	body, err := io.ReadAll(r.Body)
	if err != nil {
		errMsg := fmt.Errorf("Error reading body: %w", err)
		sendJsonMemResponse(w, 400, false, errMsg.Error())
		return
	}
	if CHATTY_SERVER {
		fmt.Printf("%s\n", body)
	}

	var memReq MemoryRequest
	err = json.Unmarshal(body, &memReq)
	if err != nil {
		errMsg := fmt.Errorf("Error parsing JSON: %w", err)
		sendJsonMemResponse(w, 400, false, errMsg.Error())
		return
	}

	decodedBuffer, err := base64.StdEncoding.DecodeString(memReq.Mem)
	if err != nil {
		errMsg := fmt.Errorf("Error decoding memory string: %w", err)
		sendJsonMemResponse(w, 400, false, errMsg.Error())
		return
	}

	updatedBuffer := simulateTick(decodedBuffer, uintptr(memReq.Offset))

	encodedMem := base64.StdEncoding.EncodeToString(updatedBuffer)
	sendJsonMemResponse(w, 200, true, encodedMem)
}
